### PR TITLE
Implement BIP66

### PR DESCRIPTION
**Pull request by sipa**
See the corresponding BIP text (see bitcoin/bips#135) and mailing list discussion for details.

### CODE REVIEW

**Comment by laanwj** [comment]
Cool, looks good to me.
I'll have a try at filing in the TODOs in the RPC test.

**Comment by sipa** [comment]
@laanwj That'd be awesome, but I don't see how to easily do that without adding functionality to the RPC framework to build its own blocks. Bitcoin Core won't mine blocks with transactions that violate the rules it wants. I guess there could be a -overridemempoolvalidationflags or something, but that gets really ugly.

**Comment by laanwj** [comment]
@sipa Bleh, yes, I just noticed; added python code to fuzz a transaction into non-canonical-DER, but it's impossible to get it into the mempool.
Building blocks from the test framework would be neat, of course.

File: qa/rpc-tests/bipdersig.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            self.nodes[2].setgenerate(True, 99)
+        self.sync_all()
+        if (self.nodes[0].getblockcount() != cnt + 1049):
+            raise AssertionError("Failed to mine 198 version=3 blocks")
+
+        # Mine 1 old-version block
+        self.nodes[1].setgenerate(True, 1)
+        self.sync_all()
+        if (self.nodes[0].getblockcount() != cnt + 1050):
+            raise AssertionFailure("Failed to mine a version=2 block after 949 version=3 blocks")
```
Comment (laanwj) [inline qa/rpc-tests/bipdersig.py]: s/AssertionFailure/AssertionError/

**Comment by laanwj** [comment]
tested ACK. Verified that the deployment behavior as mentioned in BIP66 is indeed what is enforced.

**Comment by sipa** [comment]
Addressed comments and rebased.

**Comment by luke-jr** [comment]
utACK: read over non-test commits (), looks reasonable (I did not compare to DER spec).
80ad135a5e54e8a065fee5ef36e57034679111ab
5a47811da5158df763aa2fca09ce646ee0c51e7b

**Comment by sdaftuar** [comment]
I was able to test deployment behavior, submitting regtest blocks with a couple kinds of non-DER signatures over the p2p network to a node running the new code -- worked as expected.

**Comment by fanatid** [comment]
Why BIP66 not checked that `r` or `s` less than 33 bytes?
related: https://github.com/bitcoinjs/bip66/pull/2

**Comment by vietlq** [comment]
@fanatid the example you posted should be invalidated as an invalid signature even before coming to BIP66 validation code. Layered filters I'd say, so no worries.

**Pull request merged**